dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -324,44 +323,67 @@ protected void githubWebhookDelete(String repository, String gitReference, Strin
+            try {
+                transactionHelper.transaction(() -> {
+
+                    // Retrieve the workflow
+                    Workflow workflow = workflowDAO.findById(workflowId);
+                    if (workflow == null) {
+                        LOG.info("workflow no longer exists, workflowId={}", workflowId);
+                        return;
+                    }
+
+                    // A version should be deleted if it has the same git reference name and is not frozen
+                    Predicate<Version> shouldDeleteVersion = version -> Objects.equals(version.getName(), gitReferenceName.get()) && !version.isFrozen();
 
+                    // If the default version is going to be deleted, select a new default version
+                    Version defaultVersion = workflow.getActualDefaultVersion();
+                    if (defaultVersion != null && shouldDeleteVersion.test(defaultVersion)) {
+                        Optional<WorkflowVersion> max = workflow.getWorkflowVersions().stream()
+                            .filter(v -> !Objects.equals(v.getName(), gitReferenceName.get()))
+                            .max(Comparator.comparingLong(ver -> ver.getDate().getTime()));
+                        workflow.setActualDefaultVersion(max.orElse(null));
+                    }
+
+                    // Delete all matching versions
+                    // If one or more versions were deleted, update the appropriate state
+                    if (workflow.getWorkflowVersions().removeIf(shouldDeleteVersion)) {
+                        // Update the file formats of the entry
+                        FileFormatHelper.updateEntryLevelFileFormats(workflow);
+
+                        if (workflow.getIsPublished() && workflow.getWorkflowVersions().isEmpty()) {
+                            // Unpublish the workflow if it was published and no longer has any versions
+                            User user = GitHubHelper.findUserByGitHubUsername(tokenDAO, userDAO, username, false);
+                            publishWorkflow(workflow, false, user);
+                        } else {
+                            // Otherwise, update the public state
+                            PublicStateManager.getInstance().handleIndexUpdate(workflow, StateManagerMode.UPDATE);
+                        }
+
+                        // Create a lambda event that describes the deletion
+                        LambdaEvent lambdaEvent = createBasicEvent(repository, gitReference, username, LambdaEvent.LambdaEventType.DELETE, true, deliveryId, computeWorkflowName(workflow));
+                        lambdaEventDAO.create(lambdaEvent);
+                    }
+                });
+            } catch (RuntimeException ex) {
+                LOG.error(String.format("failed to delete version, workflowId=%s repository=%s gitReference=%s", workflowId, repository, gitReference), ex);
```

Comment:
There will be no LambdaEvent for a failure. Is that intentional? I guess there wasn't one before, and things might be in such state that it's safest not to try additional DB access, but on the other hand, the user will have no idea what happened.

Also, it might be nice to know which workflows were skipped, even if only in the logs, i.e., log the remaining workflow ids that weren't processed because of this error.

Might be overkill; just throwing it out there.

Reply:
These are good ideas, but this was an optimization ticket, and we're making substantial mods to a key portion of our infra, so I tried to minimize the difference in behavior between before and after.  I'd recommend that we make any non-essential changes (like adding a LambdaEvent for failures) in a subsequent ticket.  That way, first, we can merge the current set of necessary changes, deployed them into a full complete system, and give them a good workout before we make additional tweaks that might induce/hide/compound any problems.  Adding some logging to this PR seems unlikely to break anything, though...